crates/consensus/rpc/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+#[derive(Debug)]
+pub struct RpcConsensusClient {
+    /// HTTP client to fetch blocks
+    http_client: reqwest::Client,
+    /// Handle to execution client
+    auth_server: AuthServerHandle,
+    /// Etherscan API key
+    etherscan_api_key: String,
+    /// Etherscan base API URL
+    etherscan_base_api_url: String,
+}
+
+impl RpcConsensusClient {
+    /// Create a new fake consensus client that should sent FCUs and new payloads to `auth_server`.
+    pub fn new(
+        auth_server: AuthServerHandle,
+        etherscan_api_key: String,
+        etherscan_base_api_url: String,
+    ) -> Self {
+        Self {
+            http_client: reqwest::Client::new(),
+            auth_server,
+            etherscan_api_key,
+            etherscan_base_api_url,
+        }
+    }
+
+    /// Spawn the client to start sending FCUs and new payloads by periodically fetching recent blocks.
+    pub async fn spawn<T: EngineTypes>(&self) {
+        // TODO: Add logs
+        // TODO: Generalize over block-fetching code to support different sources
+
+        let execution_client = self.auth_server.http_client();
+        let mut last_block_number: Option<u64> = None;
+
+        loop {
+            let block: EtherscanBlockResponse = self
+                .http_client
+                .get(&self.etherscan_base_api_url)
+                .query(&[
+                    ("module", "proxy"),
+                    ("action", "eth_getBlockByNumber"),
+                    ("tag", "latest"),
+                    ("boolean", "true"),
+                    ("apikey", &self.etherscan_api_key),
+                ])
+                .send()
+                .await
+                .unwrap()
+                .json()
+                .await
+                // TODO: Handle errors gracefully and do not stop the loop
+                .unwrap();
+
+            // Sleep if no new block is available
+            if block.result.header.number == last_block_number {
+                // TODO: Allow to configure this
+                sleep(Duration::from_secs(3)).await;
+                continue;
+            }
+
+            let payload = rich_block_to_execution_payload_v3(block.result);
+
+            let block_hash = payload.block_hash();
+            let block_number = payload.block_number();
+
+            // Send new events to execution client
+            reth_rpc_api::EngineApiClient::<T>::new_payload_v3(
+                &execution_client,
+                payload.execution_payload_v3,
+                payload.versioned_hashes,
+                payload.parent_beacon_block_root,
+            )
+            .await
+            .unwrap();
+            reth_rpc_api::EngineApiClient::<T>::fork_choice_updated_v3(
+                &execution_client,
+                reth_rpc_types::engine::ForkchoiceState {
+                    head_block_hash: block_hash,
+                    safe_block_hash: block_hash,
+                    finalized_block_hash: block_hash,
+                },
```

Comment:
this might break certain assumptions about blocks that can (or cannot in this case) be reorged. i'd prefer that API call above was a batch call that also fetches safe & finalized hashes

Reply:
good callout. I haven't found a way to get safe/finalized blocks state from Etherscan, because their `eth_getBlockByNumber` doesn't support tags other than `latest`. maybe you know any tricks? afaik beaconscan isn't exposed in API.

otherwise, should I just change the `--etherscan` parameter to `--consensus-rpc` and use its `eth_getBlockByNumber`?

Reply:
The safe block hash will be at most 32 blocks behind head, and finalized is at most 32 behind that, so you could use `head, head-32, head-64` after fetching the three hashes